tests/e2e/manifests/verify_manifest_lists.go

```diff
@@ -0,0 +1,674 @@
+		if i == len(s)-1 {
+			sep = "\n\n"
+		}
+		fmt.Printf(v.String() + sep)
+	}
+}
+
+func (s VersionList) Len() int {
+	return len(s)
+}
+
+func (s VersionList) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func (s VersionList) Less(i, j int) bool {
+	return s[j].LessThan(s[i])
+}
+
+func (s VersionList) Sort() {
+	sort.Sort(s)
+}
+
+// throw an error
+func exitWithError(err error) {
+	fmt.Printf("\n* ERROR: %v\n\n", err)
+	fmt.Printf(messageError)
+	os.Exit(1)
+}
+
+// prints a long line of characters
+func printLineSeparator(r rune) {
+	fmt.Println(strings.Repeat(string(r), 79))
+}
+
+// downloads the contents of a web page into a string
+func getFromURL(url string) (string, int, error) {
+	return getFromURLTimeoutSize(url, defaultHTTPTimeout, false) // use default timeout of 20 seconds
+}
+
+func getFromURLTimeoutSize(url string, timeout int, sizeOnly bool) (string, int, error) {
+	fmt.Printf("* getFromURL(): %s\n", url)
+
+	t := time.Duration(time.Duration(timeout) * time.Second)
+	client := http.Client{
+		Timeout: t,
+	}
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return "", -1, err
+	}
+	req.Header.Set("Accept", "*/*")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", -1, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return "", -1, fmt.Errorf("responed with status: %d", resp.StatusCode)
+	}
+
+	len := resp.Header.Get("Content-Length")
+	sz, err := strconv.Atoi(len)
+	if err != nil {
+		fmt.Printf("WARNING: could not covert 'Content-Length' to integer: %s", len)
+		sz = 0
+	}
+	// only the content-length is requested; do not download the whole file.
+	if sizeOnly {
+		return "", sz, nil
+	}
+
+	var src io.Reader
+	var dst bytes.Buffer
+
+	counter := &writeCounter{totalLength: int64(sz)}
+	src = io.TeeReader(resp.Body, counter)
+
+	_, err = io.Copy(&dst, src)
+	if err != nil {
+		exitWithError(err)
+	}
+	fmt.Println()
+
+	return string(dst.Bytes()), sz, nil
+}
+
+// parse the kubeadm config and obtain image versions that are not bound to the k8s version
+func getImageVersions(ver *version.Version, images map[string]string) error {
+	branch := fmt.Sprintf("release-%d.%d", ver.Major(), ver.Minor())
+	url := fmt.Sprintf("https://raw.githubusercontent.com/kubernetes/kubernetes/%s/cmd/kubeadm/app/constants/constants.go", branch)
+	constants, _, err := getFromURL(url)
+	if err != nil {
+		return err
+	}
+	lines := strings.Split(constants, "\n")
+
+	// create a map of all required images
+	k8sVersionV := "v" + ver.String()
+	images["kube-apiserver"] = k8sVersionV
+	images["kube-controller-manager"] = k8sVersionV
+	images["kube-scheduler"] = k8sVersionV
+	images["kube-proxy"] = k8sVersionV
+	images["etcd"] = ""
+	images["pause"] = ""
+	images["coredns"] = ""
+	// TODO(neolit123): kube-dns is being deprecated eventually [*]
+	images["k8s-dns-kube-dns"] = ""
+	images["k8s-dns-sidecar"] = ""
+	images["k8s-dns-dnsmasq-nanny"] = ""
+
+	// parse the constants file and fetch versions
```

Comment:
this seems a bit fragile. is there a better way to query or list the necessary packages and versions?

Reply:
sadly, no.
given the test targets kubeadm primary, there is not other location where the tags for images are defined, other than the `constants.go` file in kubeadm.